src/lib/shady-render.ts

```diff
@@ -165,30 +166,110 @@ export interface ShadyRenderOptions extends Partial<RenderOptions> {
+ * custom element's `constructor` is not supported. Instead rendering should
+ * either done asynchronously, for example at microtask timing (e.g.
+ * Promise.resolve()), or be deferred until the element's `connectedCallback`
+ * first runs.
+ *
+ * Usage considerations when using shimmed custom properties or `@apply`:
+ *
+ * * Whenever any dynamic changes are made which affect
+ * css custom properties, `ShadyCSS.styleElement(element)` must be called
+ * to update the element. This render method automatically does this on first
+ * render, but if subsequent changes are made, this call is required. When
+ * rendering via a custom element, it's common to include this call in the
+ * element's `connectedCallback`.
+ *
+ * * Rendering a ShadowRoot within a directive is not supported.
+ *
+ * * Shimmed custom properties may only be defined either for an entire
+ * shadowRoot (e.g. via `:host`) or via a rule that directly matches an element
+ * with a shadowRoot. In other words, instead of flowing from parent to child as
+ * do native css custom properties, shimmed custom properties flow only from
+ * shadowRoots to nested shadowRoots.
+ *
+ * * When using `@apply` mixing css shorthand property names with
+ * non-shorthand names (for example `border` and `border-width`) is not
+ * supported.
+ */
 export const render =
     (result: TemplateResult,
      container: Element|DocumentFragment,
      options: ShadyRenderOptions) => {
       const scopeName = options.scopeName;
       const hasRendered = parts.has(container);
+      const needsScoping = container instanceof ShadowRoot &&
+          compatibleShadyCSSVersion && result instanceof TemplateResult;
+      // Handle first render to a scope specially...
+      const firstScopeRender = needsScoping && !shadyRenderSet.has(scopeName);
+      // On first scope render, render into a fragment; this cannot be a single
+      // fragment that is reused since nested renders can occur synchronously.
+      const renderContainer =
+          firstScopeRender ? document.createDocumentFragment() : container;
       litRender(
           result,
-          container,
+          renderContainer,
           {templateFactory: shadyTemplateFactory(scopeName), ...options} as
               RenderOptions);
-      // When rendering a TemplateResult, scope the template with ShadyCSS
-      if (container instanceof ShadowRoot && compatibleShadyCSSVersion &&
-          result instanceof TemplateResult) {
-        // Scope the element template one time only for this scope.
-        if (!shadyRenderSet.has(scopeName)) {
-          const part = parts.get(container)!;
-          const instance = part.value as TemplateInstance;
+      // When performing first scope render,
+      // (1) we've rendered into a fragment so that there's a chance to
+      // `prepareTemplateStyles` before sub-elements hit the DOM
+      // (where they would normally render);
```

Comment:
...which might cause them to render based on a common pattern of rendering in connected?